erpnext/manufacturing/doctype/bom/bom.js

```diff
@@ -734,3 +740,44 @@ function trigger_process_loss_qty_prompt(frm, cdt, cdn, item_code) {
 		__("Set Quantity")
 	);
 }
+
+function add_scrap_item(frm, cdt, cdn) {
+	var item = locals[cdt][cdn];
+
+	if (item.scrap) {
+		let row = null;
+		var scrap_item_qty = (item.qty / 100) * item.scrap;
+
+		for (var i = 0; i < frm.doc.scrap_items.length; i++) {
+			if (frm.doc.scrap_items[i].item_code == item.item_code) {
+				row = frm.doc.scrap_items[i];
+				break;
+			}
+		}
+
+		if (row) {
+			row.stock_qty = scrap_item_qty;
+			row.rate = item.rate;
+			row.amount = scrap_item_qty * item.rate;
+			row.stock_uom = item.stock_uom;
+			row.base_rate = item.base_rate;
+			row.base_amount = item.base_rate * scrap_item_qty;
+		}
+		else {
+			row = frm.add_child('scrap_items', {
+				item_code: item.item_code,
+				item_name: item.item_name,
+				stock_qty: scrap_item_qty,
+				rate: item.rate,
+				amount: scrap_item_qty * item.rate,
+				stock_uom: item.stock_uom,
+				base_rate: item.base_rate,
+				base_amount: item.base_rate * scrap_item_qty,
+				is_process_loss: 0
+			});
+		}
```

Comment:
This can use lesser redundancy. All the common items can be one object, the same can update the row object . [Spread Operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) is useful here.
```suggestion
        let row_data = {
              "stock_qty" = scrap_item_qty;
			   "rate" = item.rate;
			   "amount" = scrap_item_qty * item.rate;
			   "stock_uom" = item.stock_uom;
			   "base_rate" = item.base_rate;
			   "base_amount" = item.base_rate * scrap_item_qty;
        };
        
		if (not row) {
		        row = frm.add_child('scrap_items', {
				        item_code: item.item_code,
				        item_name: item.item_name,
				        is_process_loss: 0
			    });
		}
		row = { ...row_data }; // update the existing row object with new data
```